tests/unit/models/test_model_interfaces.py

```diff
@@ -255,39 +255,37 @@ def _mock_data() -> tuple[tf.Tensor, tf.Tensor]:
 
 
 def _gpr(x: tf.Tensor, y: tf.Tensor) -> GPR:
-    return GPR((x, y), gpflow.kernels.Linear())
+    return GPR((x, y), gpflow.kernels.Matern32())
 
 
-def _sgpr(x: tf.Tensor, y: tf.Tensor) -> SGPR:
-    return SGPR((x, y), gpflow.kernels.Linear(), x[: len(x) // 2])
+def _sgpr(x: tf.Tensor, y: tf.Tensor, num_inducing: int = 2) -> SGPR:
+    return SGPR((x, y), gpflow.kernels.Matern32(), x[:num_inducing])
 
 
-def _svgp(inducing_variable: tf.Tensor) -> SVGP:
-    return SVGP(gpflow.kernels.Linear(), gpflow.likelihoods.Gaussian(), inducing_variable)
+def _svgp(x: tf.Tensor, y: tf.Tensor, num_inducing: int = 2) -> SVGP:
+    return SVGP(
+        gpflow.kernels.Matern32(), gpflow.likelihoods.Gaussian(), x[:num_inducing], num_data=len(x)
+    )
 
 
 def _vgp(x: tf.Tensor, y: tf.Tensor) -> VGP:
     likelihood = gpflow.likelihoods.Gaussian()
-    kernel = gpflow.kernels.Linear()
+    kernel = gpflow.kernels.Matern32()
     m = VGP((x, y), kernel, likelihood)
-    variational_variables = [m.q_mu.unconstrained_variable, m.q_sqrt.unconstrained_variable]
-    gpflow.optimizers.Scipy().minimize(m.training_loss_closure(), variational_variables)
     return m
 
 
 def _vgp_matern(x: tf.Tensor, y: tf.Tensor) -> VGP:
     likelihood = gpflow.likelihoods.Gaussian()
     kernel = gpflow.kernels.Matern32(lengthscales=0.2)
     m = VGP((x, y), kernel, likelihood)
-    variational_variables = [m.q_mu.unconstrained_variable, m.q_sqrt.unconstrained_variable]
-    gpflow.optimizers.Scipy().minimize(m.training_loss_closure(), variational_variables)
     return m
 
 
 class _ModelFactoryType(Protocol):
     def __call__(
         self, x: TensorType, y: TensorType, optimizer: Optimizer | None = None
-    ) -> GaussianProcessRegression:
+    ) -> tuple[GaussianProcessRegression, Callable]:
```

Comment:
I'm not sure this would be a `callable`, perhaps some parent class of `GPR`, `SVGP` etc - `GPModel` perhaps?

Reply:
If it is not an instance of the class yet (since it hasn't been initialized yet) is it a `Callable` or a `GPModel`?

Reply:
huh, not sure, I'm not an expert on typing and mypy - @uri-granta or @joelberkeley?

Reply:
it's a `Callable` **and** a `type[<whatever_model_type>]`, but I'd be generally wary of using the latter - there's no guarantee `__init__`'s signature is preserved in subclasses

that said, can you say more about why you've chosen to use SVGP to compare to SpareVariational etc? I would have thought they're all trying to approximate a GPR and that comparing to GPR would be sufficient

Reply:
also, `Callable` takes arguments e.g. `Callable[[Dataset], GaussianProcessRegression]` for input `Dataset` and output `GaussianProcessRegression`. Do use those if possible

Reply:
Just to be sure, in the latter case you're saying the correct typing would then be `type[<whatever model type>]`? I'll have a look at adding the args as you suggest, thanks!

For your second question, there's really no guarantee that SVGP will be close to GPR or any of the other model types, particularly as SVGP can be quite finicky to train. So after discussing with Henry and Hrvoje, we decided it would be best to change things so that we compare the behavior of the interfaces to the corresponding reference models. It felt like previously the tests were relying on the simplicity of the dataset and model to pass when compared to GPR, instead of actually testing whether the internal models were acting as a reference model would. So basically I changed things to 1) use more complicated data and kernel, and 2) compare the internal interface models to the equivalent reference models.

Reply:
re first qu, i've looked at the code more carefully, and I see that actually `type[<whatever model type>]` isn't right at all. It would be if you were returning the model constructors, e.g. the `SVGP` class itself is a `type[SVGP]` (as opposed to an `SVGP(...)` which is an `SVGP`. But here, we're returning the functions `_gpr` etc, which are `Callable`s. It's not possible in Python to specify the optional arg `num_inducing: int = 2` in `_svgp` etc. with `Callable`. The only way to do that is to write a `Protocol` (just like the one here) with an appropriate `__call__` method. That said, I'd first suggest it's better to have each test doing one thing rather than each of them testing all the models. If a test requires a model type like `SVGP` that has inducing points, then have a separate set of tests for those. If there is commonality between that test and those without inducing points, there may be other abstractions that can reduce duplication.

re second qu, ok, thanks. It's interesting to learn that. I'm a beginner with sparse GPs and VI and keen to learn where I can, but I was of the understanding that they're approximations to vanilla GPs. Now perhaps the approximation doesn't imply similar predictions/lml, but maybe there are other metrics that demonstrate similarity, in which case we could test for them instead. Even if this is accurate, it's likely outside the scope of this work

Reply:
re first qu, i've looked at the code more carefully, and I see that actually `type[<whatever model type>]` isn't right at all. It's not possible in Python to specify the optional arg `num_inducing: int = 2` in `_svgp` etc. with `Callable`, but it looks like you're not using that anyway - is that right? If you are, you can use a `Protocol` (just like the one here) with an appropriate `__call__` method.

regarding the return type, perhaps they're `Internal/External...Mixins` from gpflow, can't recall the name. That said, I'd ask if it's better to have each test doing one thing rather than each of them testing all the models. If a test requires an `External...` model type like `SVGP` you could have a separate set of tests for those. If there is commonality between that test and those for `Internal...`, there may be other abstractions that can reduce duplication.

Reply:
For the typing: I just had a thought and that's that I don't think we really ever set `num_inducing` to anything other than the default in these tests, and if I'm honest I don't really see a need to set them to anything else. So maybe we can just hard code it and get rid of the arg? 

That's an interesting suggestion with the `Internal` vs `External` models, and that might solve some of the problems with having to check the model type and then decide whether or not you need to pass data to train. I'll have a think about it for sure!

One of the main motivations for a bunch of these changes was that we're planning on adding an interface to GPflux for DGPs, which won't act at all like GPR models. So that's partly why we made the changes with the reference models. But maybe then it would also make much more sense to have separate tests for different model types as you suggest.

Reply:
> So maybe we can just hard code it and get rid of the arg?

that would make things easier for sure

> we're planning on adding an interface to GPflux for DGPs

ah good to hear that's in the works